category_theory/products.lean

```diff
@@ -0,0 +1,106 @@
+include 𝒞 𝒟
+
+@[simp,ematch] lemma id (X : C) (Y : D) : 𝟙 (X, Y) = (𝟙 X, 𝟙 Y) := rfl
+@[simp,ematch] lemma comp {P Q R : C} {S T U : D} (f : (P, S) ⟶ (Q, T)) (g : (Q, T) ⟶ (R, U)) : f ≫ g = (f.1 ≫ g.1, f.2 ≫ g.2) := rfl
+end
+
+section 
+variables (C : Type u₁) [𝒞 : category.{u₁ v₁} C] (D : Type u₁) [𝒟 : category.{u₁ v₁} D]
+include 𝒞 𝒟
+
+/-- 
+`product_category.uniform C D` is an additional instance specialised so both factors have the same universe levels. This helps typeclass resolution.
+-/
+instance uniform : category.{u₁ v₁} (C × D) := category_theory.product_category C D
+end
+
+-- Next we define the natural functors into and out of product categories. For now this doesn't address the universal properties.
+
+/-- `right_injection_at C Z` is the functor `X ↦ (X, Z)`. -/
+definition right_injection_at (C : Type u₁) [category.{u₁ v₁} C] {D : Type u₁} [category.{u₁ v₁} D] (Z : D) : C ↝ (C × D) := 
+{ obj      := λ X, (X, Z),
+  map      := λ X Y f, (f, 𝟙 Z),
+  map_id   := begin /- `obviously'` says: -/ intros, refl end,
+  map_comp := begin /- `obviously'` says: -/ intros, dsimp, simp end }
+
+/-- `left_injection_at Z D` is the functor `X ↦ (Z, X)`. -/
+definition left_injection_at {C : Type u₁} [category.{u₁ v₁} C] (Z : C) (D : Type u₁) [category.{u₁ v₁} D] : D ↝ (C × D) := 
+{ obj      := λ X, (Z, X),
+  map      := λ X Y f, (𝟙 Z, f),
+  map_id   := begin /- `obviously'` says: -/ intros, refl end,
+  map_comp := begin /- `obviously'` says: -/ intros, dsimp, simp end }
+
+/-- `left_projection` is the functor `(X, Y) ↦ X`. -/
+definition left_projection (C : Type u₁) [category.{u₁ v₁} C] (Z : C) (D : Type u₁) [category.{u₁ v₁} D] : (C × D) ↝ C := 
+{ obj      := λ X, X.1,
+  map      := λ X Y f, f.1,
+  map_id   := begin /- `obviously'` says: -/ intros, refl end,
+  map_comp := begin /- `obviously'` says: -/ intros, refl end }
+
+/-- `right_projection` is the functor `(X, Y) ↦ Y`. -/
+definition right_projection (C : Type u₁) [category.{u₁ v₁} C] (Z : C) (D : Type u₁) [category.{u₁ v₁} D] : (C × D) ↝ D := 
+{ obj      := λ X, X.2,
+  map      := λ X Y f, f.2,
+  map_id   := begin /- `obviously'` says: -/ intros, refl end,
+  map_comp := begin /- `obviously'` says: -/ intros, refl end }
+
+end product_category
+
+variables {A : Type u₁} [𝒜 : category.{u₁ v₁} A] {B : Type u₂} [ℬ : category.{u₂ v₂} B] {C : Type u₃} [𝒞 : category.{u₃ v₃} C] {D : Type u₄} [𝒟 : category.{u₄ v₄} D]
+include 𝒜 ℬ 𝒞 𝒟
+
+namespace functor
+/-- The cartesion product of two functors. -/
+definition product (F : A ↝ B) (G : C ↝ D) : (A × C) ↝ (B × D) :=
+{ obj := λ X, (F X.1, G X.2),
+  map := λ _ _ f, (F.map f.1, G.map f.2),
+  map_id   := begin /- `obviously'` says: -/ intros, cases X, dsimp, rw map_id_lemma, rw map_id_lemma end,
+  map_comp := begin /- `obviously'` says: -/ intros, cases Z, cases Y, cases X, cases f, cases g, dsimp at *, rw map_comp_lemma, rw map_comp_lemma end }
+
+notation F `×` G := product F G
```

Comment:
This is not going to work. Notation overloading in lean 3 is broken, so mathlib tries to minimize it. Just use `F.prod G` and `α.prod β` here. (Maybe lean 4 will improve the situation, but this is just an accident waiting to happen until then.)